sweep/fee_bumper.go

```diff
@@ -0,0 +1,623 @@
+// If so, it will bump the fee and updates the monitor record once succeeded.
+//
+// NOTE: must run as a goroutine.
+func (f *feeBumper) handleFeeBump(r *monitoredInput) {
+	defer f.wg.Done()
+
+	op := r.outpoint
+	feeLog.Debugf("Preparing fee bump for outpoint: %v", op)
+
+	newFeeRate, needUpdate := f.needFeeBump(r)
+
+	// Check if we need to do a fee bump for this tx.
+	if !needUpdate {
+		feeLog.Debugf("Skipped fee bump at height=%v for outpoint: "+
+			"%v, current fee rate: %v, estimated new fee rate: %v",
+			op, f.currentHeight, r.currentFeeRate, newFeeRate)
+
+		return
+	}
+
+	// Send the bump request to the sweeper.
+	err := f.sendFeeBump(r, newFeeRate)
+	if err != nil {
+		feeLog.Errorf("Fee bump for outpoint: %v failed, error from "+
+			"update params: %v, record removed", op, err)
+
+		f.monitoredInputs.Delete(op)
+
+		return
+	}
+
+	feeLog.Infof("Fee bump for outpoint: %v succeeded, old fee rate: %v, "+
+		"current fee rate: %v", op, r.currentFeeRate, newFeeRate)
+
+	// Update the current fee rate on this record.
+	r.currentFeeRate = newFeeRate
+
+	// Update our monitored outpoints map.
+	f.monitoredInputs.Store(op, r)
+}
+
+// sendFeeBump sends a params update request to the sweeper.
+func (f *feeBumper) sendFeeBump(r *monitoredInput,
+	newFeeRate chainfee.SatPerKWeight) error {
+
+	op := r.outpoint
+	feeLog.Debugf("Fee bumping at height=%v for outpoint: %v, current "+
+		"fee rate: %v, estimated new fee rate: %v", op,
+		f.currentHeight, r.currentFeeRate, newFeeRate)
+
+	// Note that if we bump an already confirmed transaction, an error will
+	// be returned.
+	params := ParamsUpdate{
+		Fee: FeePreference{
+			FeeRate: newFeeRate,
+		},
+		Force: true,
+	}
+	resp, err := f.updateParams(*op, params)
+	if err != nil {
+		return err
+	}
+
+	select {
+	case result := <-resp:
+		switch {
+		// We may get an ErrNotMine if the input has already been
+		// swpet.
+		case errors.Is(result.Err, lnwallet.ErrNotMine):
+			feeLog.Debugf("Outpoint: %v not found in sweeper's "+
+				"pendingInputs, assuming already spent", op)
+
+			return nil
+
+		case result.Err != nil:
+			return err
+		}
+
+		return nil
+
+	case <-f.quit:
+		feeLog.Debugf("Fee bumper shutting down, exit sendFeeBump "+
+			"for %v", op)
+		return nil
+	}
+}
+
+// monitor is the main goroutine that manages the fee bumping.
+//
+// NOTE: must run as a goroutine.
+func (f *feeBumper) monitor() {
+	defer f.wg.Done()
+
+	// Subscribe to block height updates.
+	blockEpochs, err := f.chainNotifier.RegisterBlockEpochNtfn(nil)
+	if err != nil {
+		feeLog.Errorf("register block epoch ntfn: %w", err)
+		return
+	}
+
+	for {
+		select {
+		case epoch := <-blockEpochs.Epochs:
+			height := epoch.Height
+			feeLog.Debugf("Received new block height: %v", height)
+
+			// Update our current block height.
+			f.currentHeight = uint32(height)
+
+			// Process the monitored outpoints.
+			f.monitoredInputs.Range(func(_ *wire.OutPoint,
+				r *monitoredInput) bool {
+
+				f.wg.Add(1)
```

Comment:
I wonder if this should actually be done serially 🤔

Though with the way the sweeper works, all the goroutines will be serialized anyway, so it's nice to not have to block here so we can proceed other blocks/requests.